uwrt_mars_rover_drivetrain/uwrt_mars_rover_drivetrain_description/urdf/drivetrain.macro.xacro

```diff
@@ -0,0 +1,93 @@
+
+    <!-- control parameters -->
+    <xacro:property name="mechanical_reduction" value="${control_parameters['mechanical_reduction']}" scope="parent"/>
+    <xacro:property name="max_speed" value="${control_parameters['max_speed']}" scope="parent"/>
+  </xacro:macro>
+
+  <xacro:macro name="default_inertial" params="mass:=5">
+    <inertial>
+      <mass value="${mass}" />
+      <inertia
+          ixx="1e-3" ixy="0.0" ixz="0.0"
+          iyy="1e-3" iyz="0.0"
+          izz="1e-3"/>
+    </inertial>
+  </xacro:macro>
+
+  <xacro:macro name="identical_visual_collision" params="**identical_blocks">
+    <visual>
+      <xacro:insert_block name="identical_blocks"/>
+    </visual>
+    <collision>
+      <xacro:insert_block name="identical_blocks"/>
+    </collision>
+  </xacro:macro>
+
+  <xacro:macro name="wheel" params="prefix suffix parent_link reflect wheel_diameter:=^ wheel_width:=^ wheel_mass:=^ wheelbase_length:=^ x_position" >
+    <joint name="${prefix}_${suffix}_wheel_joint" type="continuous">
+      <axis xyz="0 1 0" rpy="0 0 0" />
+      <parent link="${parent_link}"/>
+      <child link="${prefix}_${suffix}_wheel"/>
+      <origin xyz="${x_position} ${wheelbase_length/2*reflect} 0" rpy="0 0 0"/>
+    </joint>
+
+    <link name="${prefix}_${suffix}_wheel">
+      <xacro:identical_visual_collision>
+        <idential_blocks>
+          <origin xyz="0 0 0" rpy="${pi/2} 0 0" />
+          <geometry>
+            <cylinder radius="${wheel_diameter/2}" length="${wheel_width}"/>
+          </geometry>
+        </idential_blocks>
+      </xacro:identical_visual_collision>
+      <xacro:default_inertial mass="${wheel_mass}"/>
+    </link>
+  </xacro:macro>
+
+  <xacro:macro name="drivetrain_actuator" params="prefix suffix">
+    <ros2_control name="${prefix}_${suffix}_wheel_actuator" type="actuator">
+      <hardware>
+        <plugin>uwrt_mars_rover_drivetrain_hw/UWRTMarsRoverDrivetrainHWActuatorInterface</plugin>
+        <!-- TODO: CAN interface and Odrive id?-->
+      </hardware>
+      <joint name="${prefix}_${suffix}_wheel_joint">
+        <command_interface name="velocity">
+          <param name="min">${-max_speed}</param>
+          <param name="max">${max_speed}</param>
+        </command_interface>
+        <state_interface name="position"/>
+        <state_interface name="velocity"/>
+        <state_interface name="iq_current"/>
+        <!-- TODO: acceleration, torque? -->
+      </joint>
+      <transmission name="transmission1">
+        <!--  TODO: These are not actually parsed by ros2control yet: https://github.com/ros-controls/ros2_control/issues/32, https://github.com/ros-controls/ros2_control/issues/599 -->
+        <plugin>transmission_interface/SimpleTransmission</plugin>
+        <joint name="${prefix}_${suffix}_wheel_joint" role="joint1">
+          <mechanical_reduction>${mechanical_reduction}</mechanical_reduction>
+        </joint>
+      </transmission>
```

Comment:
man this xacro makes it look so immaculate and beautiful

Reply:
yup! thats the beauty of xacro. suuuuuuuper clean